diff --git a/src/Makefile.am b/src/Makefile.am
index d77c731..6918323 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -23,7 +23,7 @@ dbusintrospectdir = $(datarootdir)/sssd/introspect
 dbuspolicydir = $(sysconfdir)/dbus-1/system.d
 localedir = @localedir@
 nsslibdir = @nsslibdir@
-pamlibdir = $(nsslibdir)/security
+pamlibdir = @pammoddir@
 
 dbpath = @dbpath@
 pluginpath = @pluginpath@
@@ -273,6 +273,8 @@ SSSD_FAILOVER_OBJ = \
 SSSD_LIBS = \
     $(TALLOC_LIBS) \
     $(TEVENT_LIBS) \
+    $(TDB_LIBS) \
+    $(OPENLDAP_LIBS) \
     $(POPT_LIBS) \
     $(LDB_LIBS) \
     $(DBUS_LIBS) \
@@ -296,6 +298,8 @@ PYTHON_BINDINGS_LIBS = \
 TOOLS_LIBS = \
     $(TALLOC_LIBS) \
     $(TEVENT_LIBS) \
+    $(TDB_LIBS) \
+    $(OPENLDAP_LIBS) \
     $(POPT_LIBS) \
     $(LDB_LIBS) \
     $(DBUS_LIBS) \
@@ -961,9 +965,14 @@ if HAVE_SUSE
     dist_init_SCRIPTS += \
         sysv/SUSE/sssd
 else
+if HAVE_GENTOO
+    dist_init_SCRIPTS += \
+	    sysv/gentoo/sssd
+else
     dist_init_SCRIPTS += \
         sysv/sssd
 endif
+endif
 
 
 dist_sssdconf_DATA = \
diff --git a/src/conf_macros.m4 b/src/conf_macros.m4
index 65de7fb..aa57802 100644
--- a/src/conf_macros.m4
+++ b/src/conf_macros.m4
@@ -108,6 +108,9 @@ AC_DEFUN([WITH_INIT_DIR],
                 ]
                )
     initdir="${sysconfdir}/rc.d/init.d"
+    if test x$osname == xgentoo; then
+        initdir="${sysconfdir}/init.d"
+    fi
     if test x"$with_init_dir" != x; then
         initdir=$with_init_dir
     fi
diff --git a/src/configure.ac b/src/configure.ac
index 8ee2da4..da78a29 100644
--- a/src/configure.ac
+++ b/src/configure.ac
@@ -56,22 +56,12 @@ AC_ARG_ENABLE([nsslibdir], [AS_HELP_STRING([--enable-nsslibdir],
               [nsslibdir=$libdir])
 AC_SUBST(nsslibdir)
 
-m4_include(conf_macros.m4)
-WITH_DB_PATH
-WITH_PLUGIN_PATH
-WITH_PID_PATH
-WITH_LOG_PATH
-WITH_PUBCONF_PATH
-WITH_PIPE_PATH
-WITH_INIT_DIR
-WITH_TEST_DIR
-WITH_MANPAGES
-WITH_XML_CATALOG
-WITH_KRB5_PLUGIN_PATH
-WITH_PYTHON_BINDINGS
-WITH_SELINUX
-WITH_NSCD
-WITH_SEMANAGE
+#Set the PAM module install path
+AC_ARG_ENABLE([pammoddir], [AS_HELP_STRING([--enable-pammoddir],
+                                        [Where to install pam modules ($libdir/security)])],
+              [pammoddir=$enableval],
+              [pammoddir=$libdir/security])
+AC_SUBST(pammoddir)
 
 m4_include([external/platform.m4])
 m4_include([external/pkg.m4])
@@ -98,6 +88,23 @@ m4_include([external/libkeyutils.m4])
 m4_include([external/nsupdate.m4])
 m4_include([util/signal.m4])
 
+m4_include(conf_macros.m4)
+WITH_DB_PATH
+WITH_PLUGIN_PATH
+WITH_PID_PATH
+WITH_LOG_PATH
+WITH_PUBCONF_PATH
+WITH_PIPE_PATH
+WITH_INIT_DIR
+WITH_TEST_DIR
+WITH_MANPAGES
+WITH_XML_CATALOG
+WITH_KRB5_PLUGIN_PATH
+WITH_PYTHON_BINDINGS
+WITH_SELINUX
+WITH_NSCD
+WITH_SEMANAGE
+
 PKG_CHECK_MODULES([DBUS],[dbus-1])
 dnl if test -n "`$PKG_CONFIG --modversion dbus-1 | grep '^0\.'`" ; then
 if ! $PKG_CONFIG --atleast-version 1.0.0 dbus-1; then
diff --git a/src/external/libtevent.m4 b/src/external/libtevent.m4
index 6a0e36a..db660a2 100644
--- a/src/external/libtevent.m4
+++ b/src/external/libtevent.m4
@@ -1,7 +1,9 @@
+PKG_CHECK_MODULES(TEVENT, tevent, [], [
+    AC_CHECK_HEADER(tevent.h, [
+        AC_CHECK_LIB(tevent, tevent_context_init, [TEVENT_LIBS="-ltevent"])
+    ])
+])
+
 AC_SUBST(TEVENT_OBJ)
 AC_SUBST(TEVENT_CFLAGS)
 AC_SUBST(TEVENT_LIBS)
-
-AC_CHECK_HEADER(tevent.h,
-   [AC_CHECK_LIB(tevent, tevent_context_init, [TEVENT_LIBS="-ltevent"]) ],
-   [PKG_CHECK_MODULES(TEVENT, tevent)])
diff --git a/src/external/platform.m4 b/src/external/platform.m4
index 0bc221a..051fe07 100644
--- a/src/external/platform.m4
+++ b/src/external/platform.m4
@@ -1,11 +1,12 @@
 AC_ARG_WITH([os],
-            [AC_HELP_STRING([--with-os=OS_TYPE], [Type of your operation system (fedora|redhat|suse)])]
+            [AC_HELP_STRING([--with-os=OS_TYPE], [Type of your operation system (fedora|redhat|suse|gentoo)])]
            )
 osname=""
 if test x"$with_os" != x ; then
     if test x"$with_os" = xfedora || \
        test x"$with_os" = xredhat || \
        test x"$with_os" = xsuse || \
+       test x"$with_os" = xgentoo || \
        test x"$with_os" = xdebian ; then
         osname=$with_os
     else
@@ -22,6 +23,8 @@ if test x"$osname" = x ; then
         osname="suse"
     elif test -f /etc/debian_version ; then
         osname="debian"
+    elif test -f /etc/gentoo-release ; then
+        osname="gentoo"
     fi
 
     AC_MSG_NOTICE([Detected operating system type: $osname])
@@ -31,6 +34,7 @@ AM_CONDITIONAL([HAVE_FEDORA], [test x"$osname" = xfedora])
 AM_CONDITIONAL([HAVE_REDHAT], [test x"$osname" = xredhat])
 AM_CONDITIONAL([HAVE_SUSE], [test x"$osname" = xsuse])
 AM_CONDITIONAL([HAVE_DEBIAN], [test x"$osname" = xdebian])
+AM_CONDITIONAL([HAVE_GENTOO], [test x"$osname" = xgentoo])
 
 AC_CHECK_MEMBERS([struct ucred.pid, struct ucred.uid, struct ucred.gid], , ,
                  [[#define _GNU_SOURCE
diff --git a/src/sysv/gentoo/sssd b/src/sysv/gentoo/sssd
new file mode 100755
index 0000000..cbaedfd
--- /dev/null
+++ b/src/sysv/gentoo/sssd
@@ -0,0 +1,18 @@
+#!/sbin/runscript
+
+depend(){
+    need localmount netmount clock
+    use syslog
+}
+
+start(){
+    ebegin "Starting sssd"
+    start-stop-daemon --start --exec /usr/sbin/sssd -- -D
+    eend ${?}
+}
+
+stop(){
+    ebegin "Stopping sssd"
+    start-stop-daemon --stop --pidfile /var/run/sssd.pid
+    eend ${?}
+}
