diff --git a/src/Makefile.am b/src/Makefile.am
index d77c731..7484494 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -273,6 +273,9 @@ SSSD_FAILOVER_OBJ = \
 SSSD_LIBS = \
     $(TALLOC_LIBS) \
     $(TEVENT_LIBS) \
+    $(TDB_LIBS) \
+    $(OPENLDAP_LIBS) \
+    $(LDAP_LIBS) \
     $(POPT_LIBS) \
     $(LDB_LIBS) \
     $(DBUS_LIBS) \
@@ -296,6 +299,9 @@ PYTHON_BINDINGS_LIBS = \
 TOOLS_LIBS = \
     $(TALLOC_LIBS) \
     $(TEVENT_LIBS) \
+    $(TDB_LIBS) \
+    $(LDAP_LIBS) \
+    $(OPENLDAP_LIBS) \
     $(POPT_LIBS) \
     $(LDB_LIBS) \
     $(DBUS_LIBS) \
@@ -961,9 +967,14 @@ if HAVE_SUSE
     dist_init_SCRIPTS += \
         sysv/SUSE/sssd
 else
+if HAVE_GENTOO
+    dist_init_SCRIPTS += \
+	    sysv/gentoo/sssd
+else
     dist_init_SCRIPTS += \
         sysv/sssd
 endif
+endif
 
 
 dist_sssdconf_DATA = \
diff --git a/src/external/platform.m4 b/src/external/platform.m4
index 0bc221a..051fe07 100644
--- a/src/external/platform.m4
+++ b/src/external/platform.m4
@@ -1,11 +1,12 @@
 AC_ARG_WITH([os],
-            [AC_HELP_STRING([--with-os=OS_TYPE], [Type of your operation system (fedora|redhat|suse)])]
+            [AC_HELP_STRING([--with-os=OS_TYPE], [Type of your operation system (fedora|redhat|suse|gentoo)])]
            )
 osname=""
 if test x"$with_os" != x ; then
     if test x"$with_os" = xfedora || \
        test x"$with_os" = xredhat || \
        test x"$with_os" = xsuse || \
+       test x"$with_os" = xgentoo || \
        test x"$with_os" = xdebian ; then
         osname=$with_os
     else
@@ -22,6 +23,8 @@ if test x"$osname" = x ; then
         osname="suse"
     elif test -f /etc/debian_version ; then
         osname="debian"
+    elif test -f /etc/gentoo-release ; then
+        osname="gentoo"
     fi
 
     AC_MSG_NOTICE([Detected operating system type: $osname])
@@ -31,6 +34,7 @@ AM_CONDITIONAL([HAVE_FEDORA], [test x"$osname" = xfedora])
 AM_CONDITIONAL([HAVE_REDHAT], [test x"$osname" = xredhat])
 AM_CONDITIONAL([HAVE_SUSE], [test x"$osname" = xsuse])
 AM_CONDITIONAL([HAVE_DEBIAN], [test x"$osname" = xdebian])
+AM_CONDITIONAL([HAVE_GENTOO], [test x"$osname" = xgentoo])
 
 AC_CHECK_MEMBERS([struct ucred.pid, struct ucred.uid, struct ucred.gid], , ,
                  [[#define _GNU_SOURCE
diff --git a/src/sysv/gentoo/sssd b/src/sysv/gentoo/sssd
new file mode 100755
index 0000000..cbaedfd
--- /dev/null
+++ b/src/sysv/gentoo/sssd
@@ -0,0 +1,18 @@
+#!/sbin/runscript
+
+depend(){
+    need localmount netmount clock
+    use syslog
+}
+
+start(){
+    ebegin "Starting sssd"
+    start-stop-daemon --start --exec /usr/sbin/sssd -- -D
+    eend ${?}
+}
+
+stop(){
+    ebegin "Stopping sssd"
+    start-stop-daemon --stop --pidfile /var/run/sssd.pid
+    eend ${?}
+}
